cgen doesn't unwrap smartcasted interface value

### Describe the bug

<title>

### Reproduction Steps

```v
fn cast_interface[T, U](u U) T {
	$if U is $interface {
		if u is T {
			return u
		} else {
			panic('expected t to be ${typeof[T]().name}, got ${typeof[U]().name}')
		}
	} $else {
		$compile_error('not an interface')
	}
}

interface Foo {
	f()
}

struct Bar {}

fn (bar Bar) f() {
	println('Bar.f')
}

struct Baz {}

fn (baz Baz) f() {
	println('Baz.f')
}

fn (_ Bar) g() {
	println('secret method of Bar')
}

fn f(foo Foo, i int) {
	if i == 12341234 {
		println('secret!')
		cast_interface[Bar, Foo](foo).g()
	} else {
		println('a')
	}
}

fn main() {
	f(Bar{}, 1)
	f(Bar{}, 2)
	f(Bar{}, 12341234)
}
```

### Expected Behavior

Successful compile, and output:
```
a
a
secret!
secret method of Bar
```

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HMHD522YCHVFWG93SVZC3CFW.tmp.c:6945: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HMHD522YCHVFWG93SVZC3CFW.tmp.c:13074: error: cannot convert 'struct main__Foo' to 'struct main__Bar'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.4 d744314

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 57a33c3.d744314
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-01-19 17:38:37

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: 0.4.4-80-gd744314b
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4630,7 +4630,7 @@ fn (mut g Gen) ident(node ast.Ident) {
 				g.write('(*(')
 			}
 			if node.obj.smartcasts.len > 0 {
-				obj_sym := g.table.sym(node.obj.typ)
+				obj_sym := g.table.sym(g.unwrap_generic(node.obj.typ))
 				if !prevent_sum_type_unwrapping_once {
 					for _, typ in node.obj.smartcasts {
 						is_option_unwrap := is_option && typ == node.obj.typ.clear_flag(.option)
@@ -4646,7 +4646,9 @@ fn (mut g Gen) ident(node ast.Ident) {
 								g.write('*')
 							}
 						} else if (g.inside_interface_deref && g.table.is_interface_var(node.obj))
-							|| node.obj.ct_type_var == .smartcast {
+							|| node.obj.ct_type_var == .smartcast
+							|| (obj_sym.kind == .interface_
+							&& g.table.type_kind(node.obj.typ) == .any) {
 							g.write('*')
 						} else if is_option {
 							g.write('*(${g.base_type(node.obj.typ)}*)')

--- a/vlib/v/tests/generic_smartcast_test.v
+++ b/vlib/v/tests/generic_smartcast_test.v
@@ -1,0 +1,46 @@
+fn cast_interface[T, U](u U) T {
+	$if U is $interface {
+		if u is T {
+			return u
+		} else {
+			panic('expected t to be ${typeof[T]().name}, got ${typeof[U]().name}')
+		}
+	} $else {
+		$compile_error('not an interface')
+	}
+}
+
+interface Foo {
+	f()
+}
+
+struct Bar {}
+
+fn (bar Bar) f() {}
+
+struct Baz {}
+
+fn (baz Baz) f() {}
+
+fn (_ Bar) g() int {
+	return 0
+}
+
+fn (_ Baz) g() int {
+	return 1
+}
+
+fn f(foo Foo) int {
+	if foo is Bar {
+		return cast_interface[Bar, Foo](foo).g()
+	}
+	if foo is Baz {
+		return cast_interface[Baz, Foo](foo).g()
+	}
+	return -1
+}
+
+fn test_main() {
+	assert f(Bar{}) == 0
+	assert f(Baz{}) == 1
+}

